### ISSUE TITLE
Socket hang up

### ISSUE THREAD

**Issue by highfeed** [bitinn/node-fetch#232]
When I sent 30 requests that a few fall with this error. How to fix?

```
FetchError: request to https://api.telegram.org/botTOKEN/sendDocument failed, reason: socket hang up
at ClientRequest.<anonymous> (/home/PROJECT/node_modules/node-fetch/index.js:133:11)
at emitOne (events.js:96:13)
at ClientRequest.emit (events.js:188:7)
at TLSSocket.socketOnEnd (_http_client.js:346:9)
at emitNone (events.js:91:20)
at TLSSocket.emit (events.js:185:7)
at endReadableNT (_stream_readable.js:974:12)
at _combinedTickCallback (internal/process/next_tick.js:74:11)
at process._tickDomainCallback (internal/process/next_tick.js:122:9)
```

**Comment by TimothyGu** [comment]
Is this issue specific to node-fetch? Or is it reproducible with other modules like `http.request` as well?

**Comment by highfeed** [comment]
node-fetch — https://github.com/telegraf/telegraf/blob/develop/lib/network/client.js#L58

**Comment by bitinn** [comment]
I would say "socket hang up" is seldom the client's problem. However, if you are able to get it consistently, maybe you are missing some parameters that's required by Telegram.

But either way, I don't think the problem is fixable by us, because "socket hang up" is from Node.js core, and mostly due to networking issue.

**Comment by highfeed** [comment]
OK. Thanks!

**Comment by caub** [comment]
I had this error, while doing a `await fetchHeaders('/user/' + user.id, { body: { login: 'uuu' } })`, I forgot `method: 'PATCH'`

**Comment by fauna-brecht** [comment]
I'd love to add some extra information. Many users report these socket hangups periodically with node when using the Fauna JavaScript driver https://github.com/fauna/faunadb-js which seems to trace back to node-fetch. The request never arrives. It seems to be somehow correlated with either:
- Multiple parallel requests
- Somehow printing to the console while doing requests
Another peculiarity is that I was able to consistently trigger the error at a certain moment, so I tried to instrument the code to first send the request to fiddler and get some insight in what happens. However, the error went away when I instrumented the code and the call worked fine. When I removed the instrumentation the call resulted in a socket hang up. Similar calls with the exact same signature but different body and different processing going on inside node work fine. 

Would the recommendation be to use a retry similar to what the folks at gatsby did? https://www.npmjs.com/package/fetch-retry. Does it make sense that these socket hangups happen more frequently when node is processing/logging a lot?
